src/main.rs

```diff
@@ -139,204 +174,209 @@ impl DiemDemo {
-                    },
-                    Err(_) => {
-                        self.sent_events = None;
-                    }
-                }
 
                 // Init received events
                 let mut batch = JsonRpcBatch::new();
                 let received_events_key = account_view.received_events_key.0.clone();
                 batch.add_get_events_request(received_events_key.to_string(), 0, account_view.sequence_number.clone());
-                let responses = self.rpc_client.execute(batch).unwrap();
-                match get_response_from_batch(0, &responses).unwrap().as_ref() {
-                    Ok(resp) => {
-                        self.received_events = Option::from(EventView::vec_from_response(resp.clone()).unwrap());
-                    },
-                    Err(_) => {
-                        self.received_events = None;
+                if let Ok(resp) = self.request_rpc(batch) {
+                    let received_events = EventView::vec_from_response(resp.clone()).unwrap();
+                    let mut new_events: Vec<EventView> = Vec::new();
+                    for event in received_events.clone() {
+                        let exist = self.received_events.as_ref().is_some()
+                            && self.received_events.as_ref().unwrap().iter().any(|x| x.transaction_version == event.transaction_version);
+                        if !exist {
+                            println!("new received event!");
+                            new_events.push(event);
+                        }
+                    }
+
+                    if new_events.len() > 0 && !state_initiated {
+                        if let Err(_) = self.init_state(None).await {
+                            return Err(Error::FailedToInitState);
+                        }
+
+                        state_initiated = true;
+                    }
+
+                    for event in new_events {
+                        if let Ok(transaction) = self.get_transaction_by_version(event.transaction_version) {
+                            println!("received transaction:{:?}", transaction);
+                            if let Ok(transaction_with_proof) = self.get_transaction_proof(&transaction) {
+                                println!("transaction_with_proof:{:?}", transaction_with_proof);
+
+                                let transaction_with_proof_b64 = base64::encode(&bcs::to_bytes(&transaction_with_proof).unwrap());
+                                let _resp = pr.query(DIEM_CONTRACT_ID, QueryReqData::VerifyTransaction
+                                    { account_address: account_address.clone(), transaction_with_proof_b64 }).await?;
+                            } else {
+                                println!("get_transaction_proof error");
+                            }
+                        } else {
+                            println!("get_transaction_by_version error");
+                        }
```

Comment:
```suggestion
                        let transaction = self.get_transaction_by_version(event.transaction_version)?;
                        println!("received transaction:{:?}", transaction);
                        
                        let Ok(transaction_with_proof) = self.get_transaction_proof(&transaction)?;
                        println!("transaction_with_proof:{:?}", transaction_with_proof);

                        let transaction_with_proof_b64 = base64::encode(&bcs::to_bytes(&transaction_with_proof).unwrap());
                        pr.query(DIEM_CONTRACT_ID, QueryReqData::VerifyTransaction
                                    { account_address: account_address.clone(), transaction_with_proof_b64 }).await?;
```

Reply:
Use `?` to handle errors. Also when error occurs we can do nothing rather than crash (or rerun the outer loop)